### ISSUE TITLE
Duplicate param name

### ISSUE THREAD

**Issue by Cherry** [honojs/hono#1542]
### What version of Hono are you using?

3.7.3

### What runtime/platform is your app running on?

Cloudflare Workers

### What steps can reproduce the bug?

The following code doesn't work as expected:

```javascript
import { Hono } from 'hono'

const app = new Hono()

app.get('/:type/:url', (c) => {
	c.text(`type: ${c.req.param('type')}, url: ${c.req.param('url')}`)
});

app.get('/foo/:type/:url', (c) => {
	c.text(`foo type: ${c.req.param('type')}, url: ${c.req.param('url')}`)
});

export default app
```
It produces the following error:

```
Error: Duplicate param name, use another name instead of 'type' - GET /foo/:type/:url <--- 'type'
```

### What is the expected behavior?

Compared to a similar `express` app in Node.js:

```javascript
const express = require('express');
const app = express();
const port = 3000;

app.get('/:type/:url', (req, res) => {
  res.send(`type: ${req.params.type}, url: ${req.params.url}!`);
});

app.get('/foo/:type/:url', (req, res) => {
	res.send(`foo type: ${req.params.type}, url: ${req.params.url}!`);
});

app.listen(port, () => {
  console.log(`Example app listening on port ${port}`);
});
```

Or an `itty-router` app for Cloudflare Workers:
```javascript
import {Router, error, json} from 'itty-router'

const router = Router()

router.get('/:type/:url', (req) => {
	const params = req.params;
	return new Response(`type: ${req.params.type}, url: ${req.params.url}`)
});

router.get('/foo/:type/:url', (req) => {
	const params = req.params;
	return new Response(`foo type: ${req.params.type}, url: ${req.params.url}`)
});

export default {
	fetch: (request, ...args) => router.handle(request, ...args).then(json).catch(error)
}
```

These work just fine. This should also work fine with Hono.

### Additional information

Looking at https://github.com/honojs/hono/issues/569, it seems like this might be intentional. This isn't ideal and most other routers I've used handle this just fine. It's especially a hit to DX when trying to do things like re-use validation like this:
```javascript
import { Hono } from 'hono'
import { zValidator } from '@hono/zod-validator'
import { z } from 'zod';

const app = new Hono()

const paramsSchema = z.object({
	type: z.enum(['internal', 'external']),
	url: z.string().url().max(2048),
});

app.get('/:type/:url', zValidator('param', paramsSchema), (c) => {
	const params = c.req.valid('param');
	c.text(`type: ${params.type}, url: ${params.url}`)
});

app.get('/foo/:type/:url', zValidator('param', paramsSchema), (c) => {
	const params = c.req.valid('param');
	c.text(`foo type: ${params.type}, url: ${params.url}`)
});

export default app
```

**Comment by NicoPlyley** [comment]
@Cherry 

I know this may not be the solution you're looking for, but if you do it this way, it will work just fine

```ts
app.get('/foo/:type/:url', (c) => {
  return c.text(`foo type: ${c.req.param('type')}, url: ${c.req.param('url')}`);
});

app.get('/bar/:type/:url', (c) => {
  return c.text(`bar type: ${c.req.param('type')}, url: ${c.req.param('url')}`);
});
```
Even though that was stated in the issue you mention it does not work, it works fine for me and it's how I have my entire project setup

**Comment by Cherry** [comment]
Thanks @NicoPlyley. Unfortunately though, that doesn't result in the same thing.

With the express and itty-router examples, the following all work fine:

`/something-random/abc`
`/something-else/def`
`/foo/something-random/abc`
`/foo/something-else/def`

And I can replace `something-random` and `something-else` with any string and it works.

Unfortunately that's not the case with the hono example.

**Comment by yusukebe** [comment]
This looks like a bug, but it is expected behavior and is not a bug. I will comment on this in more detail later, but I will remove the label "bug".

Thanks.

**Comment by Cherry** [comment]
Given how this works with all other routers I’ve tried, including itty and express, this seems like very undesirable behaviour.

**Comment by isaac-mcfadyen** [comment]
+1, I'm a bit confused why this isn't considered a bug as well. I have multiple routes:
`GET /domains/:hostname`
`POST /domains/:hostname`
`PUT /domains/:hostname/configuration`

In all of the routes, I want to refer to the `hostname` of the domain. It seems counterintuitive and a bit frustrating to need to call them `hostname1`, `hostname2`, `hostname3` since they're on separate routes 🤔

**Comment by yusukebe** [comment]
It is the intended behavior. But you have a good point and I would like to support it. Please wait a while.

**Comment by yusukebe** [comment]
Hi @Cherry,

We have started working on resolving this issue. It will be a big change, so please bear with us for a bit longer. Thanks.

**Comment by Cherry** [comment]
Thanks for the update! Excited to see this progressing. ❤

**Comment by yusukebe** [comment]
This issue was fixed by #1566.

**Comment by BoBberAdmin** [comment]
Just ran into this "problem". 
Was running an older version.
Updating "fixed" it, back in business. 
Thank you!
